File: replica/src/main/java/com/lab/replica/service/ReplicaStore.java
package com.lab.replica.service;

import com.lab.replica.model.ValueVersion;
import org.springframework.stereotype.Component;

import java.util.concurrent.locks.ReentrantLock;

/**
 * Thread-safe local storage for a single register (value + version).
 */
@Component
public class ReplicaStore {

    private int value = 0;
    private int version = 0;
    private final ReentrantLock lock = new ReentrantLock();

    public ValueVersion get() {
        lock.lock();
        try {
            return new ValueVersion(value, version);
        } finally {
            lock.unlock();
        }
    }

    /**
     * Apply update only if incoming version >= current version.
     */
    public void applyIfNewer(int newValue, int newVersion) {
        lock.lock();
        try {
            if (newVersion >= this.version) {
                this.value = newValue;
                this.version = newVersion;
            }
        } finally {
            lock.unlock();
        }
    }
}

File: replica/src/main/kotlin/com/lab/replica/service/ReplicaStore.kt
package com.lab.replica.service

import com.lab.replica.model.ValueVersion
import org.springframework.stereotype.Component
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Thread-safe local register: one (value, version) pair.
 */
@Component
class ReplicaStore {

    private var current = ValueVersion(0, 0)
    private val lock = ReentrantLock()

    fun get(): ValueVersion = lock.withLock { current }

    /** Persist only if incoming version is >= current (last-write-wins). */
    fun applyIfNewer(value: Int, version: Int) = lock.withLock {
        if (version >= current.version) {
            current = ValueVersion(value, version)
        }
    }
}
